File: Python3/1003.py
class node: #也可以不用自定义 换成元组
    def __init__(self,to,dis):
        self.to = to
        self.dis = dis
    def __lt__(self, other): #重写<
        return self.dis < other.dis
from queue import PriorityQueue
n,m,c1,c2 = map(int,input().split())
weight = list(map(int,input().split()))
d = [1e5 for i in range(n)]
vis = [False for i in range(n)]
cnt = [1 for i in range(n)]
w = [0 for i in range(n)]
graph = {}
q = PriorityQueue()
for _ in range(m):
    a,b,c = map(int,input().split())
    if graph.get(a) is None:
        graph[a] = []
    graph[a].append([b,c])
    if graph.get(b) is None:
        graph[b] = []
    graph[b].append([a,c])
def Disjktra(s):
    d[s] = 0
    w[s] = weight[s]
    q.put(node(s,0))
    while q.qsize():
        cur = q.get()
        if vis[cur.to]:
            continue
        vis[cur.to] = True
        for each in graph[cur.to]:
            to = each[0]
            dis = each[1]
            if vis[to] == False and dis + cur.dis < d[to]:
                d[to] = dis + cur.dis
                w[to] = w[cur.to] + weight[to]
                cnt[to] = cnt[cur.to]
                q.put(node(to,d[to]))
            elif vis[to] == False and dis + cur.dis == d[to]:
                if w[to] < w[cur.to] + weight[to]:
                    w[to] = w[cur.to] + weight[to]
                    q.put(node(to,d[to]))
                cnt[to] += cnt[cur.to]
Disjktra(c1)
print("%d %d"%(cnt[c2],w[c2]))

File: c++/1003.cpp
//
//  main.cpp
//  test
//
//  Created by 李寻欢 on 2019/9/28.
//  Copyright © 2019 李寻欢. All rights reserved.
//
#include <stdio.h>
#include <iostream>
#include <string>
#include <map>
#include <cmath>
#include <algorithm>
#include <stack>
#include <vector>
#include <queue>
#include <unordered_map>
#include <string.h>
using namespace std;

const int INF = 1e6;
const int MAXN = 510;
int n,m,c1,c2;
typedef pair<int, int> pp;
int people[MAXN];
int d[MAXN];
int w[MAXN];
bool vis[MAXN] = { false };
int path[MAXN];
vector<pp> graph[MAXN];

void Dijkstra(int s){
    fill(path, path+MAXN, 1);
    fill(w, w+MAXN, 0);
    fill(d, d+MAXN, INF);
    priority_queue<pp,vector<pp>,greater<pp>> q;
    d[s] = 0;
    w[s] = people[s];
    q.push({0,s});
    while (!q.empty()) {
        pp cur = q.top();
        q.pop();
        if(vis[cur.second]) continue;
        vis[cur.second] = true;
        for(int i=0;i<graph[cur.second].size();++i){
            int dis = graph[cur.second][i].first;
            int to = graph[cur.second][i].second;
            if(vis[to] == false && dis + cur.first < d[to]){
                d[to] = dis + cur.first;
                w[to] = w[cur.second]+people[to];
                path[to] = path[cur.second];
                q.push({d[to],to});
            }else if (vis[to] == false && dis+cur.first == d[to]){
                if(w[to] < w[cur.second] + people[to]){
                    w[to] = w[cur.second] + people[to];
                    q.push({d[to],to});
                }
                path[to] += path[cur.second];
            }
        }
    }
}

int main(){
    scanf("%d%d%d%d",&n,&m,&c1,&c2);
    for(int i=0;i<n;++i) scanf("%d",&people[i]);
    for(int i=0;i<m;++i){
        int a,b,c;
        scanf("%d%d%d",&a,&b,&c);
        graph[a].push_back(pp(c,b));
        graph[b].push_back(pp(c,a));
    }
    Dijkstra(c1);
    printf("%d %d",path[c2],w[c2]);
    
}
